NAS-139137 / 26.04 / Add fh_to_parent export definition

anodos325 opened the pull request:
This commit adds support for converting a file handle to its parent dentry. This is called in exportfs_decode_fh_raw() when subtree checking is enabled in the NFS server configuratoin. Defining this and handling the expanded filehandles allows the knfsd to succeed in handling the file handle where it might otherwise fail with ESTALE when trying to open by filehandle.

A side effect of this change is that name_to_handle_at(2) and open_by_handle_at(2) now support AT_HANDLE_CONNECTABLE on kernel 6.13 and later.

This allows in principle an avenue to convert NFS handles into paths.

anodos325 committed:
Add fh_to_parent export definition

This commit adds support for converting a file handle to its
parent dentry. This is called in exportfs_decode_fh_raw()
when subtree checking is enabled in NFS. Defining this and
handling the expanded filehandles allows the knfsd to succeed
in handling the file handle where it might otherwise fail
with ESTALE when trying to open by filehandle.

A side effect of this change is that name_to_handle_at(2)
and open_by_handle_at(2) now support AT_HANDLE_CONNECTABLE.

Signed-off-by: Andrew Walker <andrew.walker@truenas.com>

bugclerk commented:
Jira URL: https://ixsystems.atlassian.net/browse/NAS-139137

ixhamza commented on module/os/linux/zfs/zpl_export.c:
```diff
@@ -56,11 +56,43 @@ zpl_encode_fh(struct inode *ip, __u32 *fh, int *max_len, struct inode *parent)
+		if ((rc == 0) && (len_bytes > required_len)) {
+			parent_len = len_bytes - required_len;
+			pfid = (fid_t *)((char *)fh + required_len);
+			pfid->fid_len = parent_len - offsetof(fid_t, fid_data);
```
I guess we can have a case where `parent_len < offsetof(fid_t, fid_data)` and the subtraction results in an unexpected `pfid->fid_len value`.

anodos325 committed:
Fix buffer length check

This commit fixes the check for buffer length for the parent
fhandle when encoding a file handle.

Signed-off-by: Andrew Walker <andrew.walker@truenas.com>

bugclerk commented:
JIRA ticket https://ixsystems.atlassian.net/browse/NAS-139137 is targeted to the following versions which have not received their corresponding PRs: 25.10.2

bugclerk commented:
This PR has been merged and conversations have been locked.
If you would like to discuss more about this issue please use our forums or raise a Jira ticket.
